File: csharp/UIGeneration/OverlappingPanelsNode.cs
using System;
using System.Drawing;
using System.Collections;

using PUC.Layout;

namespace PUC.UIGeneration
{
	/// <summary>
	/// This object represents nodes that branch to multiple 
	/// non-overlapping panels.
	/// </summary>
	public class OverlappingPanelsNode : MultiplePanelNode
	{
		/*
		 * Member Variables
		 */

		/*
		 * Constructor
		 */

		public OverlappingPanelsNode() : base()
		{
		}


		/*
		 * InterfaceNode Methods
		 */

		public override void CalculateMinimumSize(LayoutVariables vars)
		{
			IEnumerator e = _panels.GetEnumerator();
			
			_minSize = new System.Drawing.Size( 0, 0 );

			while( e.MoveNext() )
			{
				((InterfaceNode)e.Current).CalculateMinimumSize( vars );

				System.Drawing.Size size = ((InterfaceNode)e.Current).MinimumSize;

				_minSize.Width = Math.Max( size.Width, _minSize.Width );
				_minSize.Height = Math.Max( size.Height, _minSize.Height );
			}
		}

		public override void CalculatePreferredSize(LayoutVariables vars)
		{
			IEnumerator e = _panels.GetEnumerator();
			
			int prefWidth = 0;
			int prefHeight = 0;

			while( e.MoveNext() )
			{
				((InterfaceNode)e.Current).CalculatePreferredSize( vars );

				PreferredSize size = ((InterfaceNode)e.Current).PreferredSize;

				if ( prefWidth == PreferredSize.INFINITE || size.Width == PreferredSize.INFINITE )
					prefWidth = PreferredSize.INFINITE;
				else
					prefWidth = Math.Max( size.Width, prefWidth );

				if ( prefHeight == PreferredSize.INFINITE || size.Height == PreferredSize.INFINITE )
					prefHeight = PreferredSize.INFINITE;
				else
					prefHeight = Math.Max( size.Height, prefHeight );
			}

			_prefSize = new PUC.Layout.PreferredSize( prefWidth, prefHeight );
		}

		public override void AddComponents( PUC.CIO.ContainerCIO container, LayoutVariables vars )
		{
			Container = container;

			IEnumerator e = _panels.GetEnumerator();
			while( e.MoveNext() )
				((InterfaceNode)e.Current).AddComponents( container, vars );

			this.CalculateMinimumSize( vars );
			this.CalculatePreferredSize( vars );
		}

		public override void DoLayout( LayoutVariables vars )
		{
			IEnumerator e = _panels.GetEnumerator();
			
			while( e.MoveNext() )
			{
				InterfaceNode node = (InterfaceNode)e.Current;

				node.SetSize( _bounds.Width, _bounds.Height );
				node.SetLocation( _bounds.X, _bounds.Y );

				node.DoLayout( vars );
			}
		}

		public override string ToString()
		{
			return "OverlappingPanelsNode - " + base.ToString();
		}

	}
}

File: edu/cmu/hcii/puc/uigen/OverlappingPanelsNode.java
/**
 * OverlappingPanelsNode.java
 *
 * This node represents nodes that branch to multiple panels.
 *
 * @author Jeffrey Nichols
 */

// Package Definition

package edu.cmu.hcii.puc.uigen;


// Import Declarations

import java.awt.Dimension;
import java.awt.Rectangle;

import java.util.Enumeration;
import java.util.Vector;

import edu.cmu.hcii.puc.cio.ContainerCIO;


// Class Definition

public class OverlappingPanelsNode extends MultiplePanelNode {

    //**************************
    // Member Variables
    //**************************

    //**************************
    // Constructor
    //**************************

    public OverlappingPanelsNode() {
	super();
    }

    //**************************
    // Member Methods
    //**************************

    //**************************
    // Interface Methods
    //**************************

    public Dimension getPreferredSize( LayoutVariables pVars ) {
	Dimension ret = new Dimension();
	
	Enumeration e = m_vPanels.elements();
	while( e.hasMoreElements() ) {
	    Dimension d = ((InterfaceNode)e.nextElement()).getPreferredSize( pVars );
	    
	    if ( d.width > ret.width )
		ret.width = d.width;

	    if ( d.height > ret.height )
		ret.height = d.height;
	}

	return ret;
    }

    public void addComponents( ContainerCIO pContainer ) {

	Enumeration e = m_vPanels.elements();
	while( e.hasMoreElements() )
	    ((InterfaceNode)e.nextElement()).addComponents( pContainer );
    }

    public void doSizing( LayoutVariables pVars ) {

	Enumeration e = m_vPanels.elements();
	while( e.hasMoreElements() ) {
	    InterfaceNode pNode = (InterfaceNode)e.nextElement();

	    Rectangle rBounds = (Rectangle)m_rBounds.clone();

	    if (! pVars.m_bConstrainPanelsToScreenH ) {
		rBounds.width = pNode.getPreferredSize( pVars ).width;
	    }
	    if (! pVars.m_bConstrainPanelsToScreenV ) {
		rBounds.height = pNode.getPreferredSize( pVars ).height;
	    }

	    pNode.setSize( rBounds.width,
			   rBounds.height );
	    pNode.setLocation( rBounds.x,
			       rBounds.y );

	    pNode.doSizing( pVars );
	}
    }

    public void doLayout( ContainerCIO pContainer, 
			  LayoutVariables pVars ) {
	
	Enumeration e = m_vPanels.elements();
	while( e.hasMoreElements() ) {
	    InterfaceNode pNode = (InterfaceNode)e.nextElement();

	    pNode.doLayout( pContainer, pVars );
	}
    }

    public String toString() {

	return "OverlappingPanelsNode - " + super.toString();
    }
}
